blacksheep/testing/simulator.py

```diff
@@ -0,0 +1,79 @@
+import abc
+import json
+from typing import Any, Dict, Optional
+from urllib import parse
+
+from blacksheep.contents import ASGIContent
+from blacksheep.messages import Request
+from blacksheep.server.responses import Response
+from blacksheep.testing.helpers import get_example_scope
+from blacksheep.testing.messages import MockReceive
+
+
+def _create_scope(
+    method: str,
+    path: str,
+    headers: Optional[Dict[str, str]] = None,
+    query: Optional[Dict[str, str]] = b"",
+) -> Dict:
+    if headers is not None:
+        headers = [(key.encode(), value.encode()) for key, value in headers.items()]
+
+    if query:
+        query = parse.urlencode(query).encode()
+
+    scope = get_example_scope(method, path, extra_headers=headers, query=query)
+    return scope
+
+
+class AbstractTestSimulator:
+    @abc.abstractmethod
+    async def send_request(
+        self,
+        method: str,
+        path: str,
+        headers: Optional[Dict[str, str]] = None,
+        query: Optional[Dict[str, str]] = b"",
+        content: Dict[str, Any] = None,
+    ):
+        pass
+
+
+class TestSimulator(AbstractTestSimulator):
+    def __init__(self, app):
+        self.app = app
+        self._prepare_application()
+
+    async def send_request(
+        self,
+        method: str,
+        path: str,
+        headers: Optional[Dict[str, str]] = None,
+        query: Optional[Dict[str, str]] = b"",
+        content: Optional[Dict[str, Any]] = None,
+    ) -> Response:
+
+        scope = _create_scope(method, path, headers, query)
+        request = Request.incoming(
+            scope["method"],
+            scope["raw_path"],
+            scope["query_string"],
+            scope["headers"],
+        )
+
+        request.scope = scope
+
+        if content is not None:
+            encoded_content = json.dumps(content).encode()
+            request.content = ASGIContent(MockReceive([encoded_content]))
```

Comment:
@RobertoPrevato If we want to support JSON/multipart body in the request, we should probably create a class that will set proper content with the proper class.

e.g
```
from blacksheep.testing import Content

response = await client.post("/users", content=Content(mime_type="application/json", data={"username": "foo"}))
```

That's was my first thought, but not sure if this the best way.

Reply:
There is a `JSONContent` class, that can be used for responses and requests in the client. The `ASGIContent` class was added when I modified the framework to be ASGI compatible, it should be needed only for incoming content from the ASGI web server.

Can you please try with the existing `JSONContent` class? It should work.

Reply:
Yeah, will do.

Reply:
It just occurred to me that maybe it would be best if the `send_request` method accepted as input an `Optional[Content]`, and not assume that the user wants to handle JSON. For example it could be also Form content, Multipart, raw bytes.

Otherwise, the method signature might support different kinds of keywords (this is the approach of `requests` library), e.g. 

```python
def example(..., json=None, form=None, files=None, multipart=None, raw=None): ...
```

See: https://docs.python-requests.org/en/master/user/quickstart/#make-a-request

![image](https://user-images.githubusercontent.com/2576032/119880451-8f629180-bf2c-11eb-8294-1abf3aa29fea.png)

Reply:
What do you think?

Reply:
Yeah, I already pushed changes, and now for content argument we're expecting a Content's child class.

e.g

`JSONContent, FormContent`, etc.